presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java

```diff
@@ -1171,6 +1171,20 @@ public boolean usesLegacyTableLayouts(Session session, TableHandle table)
         return getMetadata(session, table.getConnectorId()).usesLegacyTableLayouts();
     }
 
+    @Override
+    public Optional<TableHandle> applyLimit(Session session, TableHandle table, long limit)
+    {
+        ConnectorId connectorId = table.getConnectorId();
+        ConnectorMetadata metadata = getMetadata(session, connectorId);
+
+        if (metadata.usesLegacyTableLayouts()) {
+            return Optional.empty();
+        }
+
+        return metadata.applyLimit(table.getConnectorHandle(), limit)
+                .map(newTable -> new TableHandle(connectorId, newTable, table.getTransaction(), table.getLayout()));
```

Comment:
If `table.getLayout()` is always null here, maybe just pass null, to keep the call count down.